emba.sh

```diff
@@ -385,30 +432,72 @@ main()
         architecture_dep_check
       fi
 
-      if [[ -n "${#ROOT_PATH[@]}" ]]; then
+
+      if [[ -d "$LOG_DIR" ]]; then
+        print_output "[!] Testing phase started on ""$(date)""\\n""$(indent "$NC""Firmware path: ""$FIRMWARE_PATH")" "main" 
+      else
+        print_output "[!] Testing phase started on ""$(date)""\\n""$(indent "$NC""Firmware path: ""$FIRMWARE_PATH")" "no_log"
+      fi
+      write_grep_log "$(date)" "TIMESTAMP"
+
+      if [[ "${#ROOT_PATH[@]}" -eq 0 ]]; then
         detect_root_dir_helper "$FIRMWARE_PATH"
       fi
 
       check_firmware
-
       prepare_binary_arr
+      prepare_file_arr
       set_etc_paths
       echo
 
-      print_output "[!] Test started on ""$(date)""\\n""$(indent "$NC""Firmware path: ""$FIRMWARE_PATH")" "no_log"
-      write_grep_log "$(date)" "TIMESTAMP"
-
       # 'main' functions of imported modules
 
+      # in threaded mode we start the long running modules first 
+      if [[ $THREADED -eq 1 ]]; then
+        if [[ $BAP -eq 1 ]]; then
+          MODULE_FILE="$MOD_DIR"/S120_cwe_checker.sh
+
+          MODULE_BN=$(basename "$MODULE_FILE")
+          MODULE_MAIN=${MODULE_BN%.*}
+          module_start_log "$MODULE_MAIN"
+          HTML_REPORT=0
+          $MODULE_MAIN &
+          WAIT_PIDS+=( "$!" )
+        fi
+
+        if [[ $QEMULATION -eq 1 ]]; then
+          MODULE_FILE="$MOD_DIR"/S115_usermode_emulator.sh
+          MODULE_BN=$(basename "$MODULE_FILE")
+          MODULE_MAIN=${MODULE_BN%.*}
+          module_start_log "$MODULE_MAIN"
+          HTML_REPORT=0
+          $MODULE_MAIN &
+          WAIT_PIDS+=( "$!" )
+        fi
+      fi
+
       if [[ ${#SELECT_MODULES[@]} -eq 0 ]] ; then
         local MODULES
-        mapfile -t MODULES < <(find "$MOD_DIR" -name "S*_*.sh" | sort -V 2> /dev/null)
+        if [[ $THREADED -eq 1 ]]; then
+          mapfile -t MODULES < <(find "$MOD_DIR" -name "S*_*.sh" | grep -v "emulator\|cwe" | sort -V 2> /dev/null)
+        else
+          mapfile -t MODULES < <(find "$MOD_DIR" -name "S*_*.sh" | sort -V 2> /dev/null)
+        fi
         for MODULE_FILE in "${MODULES[@]}" ; do
           if ( file "$MODULE_FILE" | grep -q "shell script" ) && ! [[ "$MODULE_FILE" =~ \ |\' ]] ; then
             MODULE_BN=$(basename "$MODULE_FILE")
             MODULE_MAIN=${MODULE_BN%.*}
+            module_start_log "$MODULE_MAIN"
```

Comment:
I do not think the log file should be different in threading vs non threading. Otherwise you have to use different logs to find out which module is already finished or not.

Reply:
Whats about that: Logfiles with start and end stamp and emba output on terminal without this information?

Reply:
like this:
echo -e "[*] $(date) - $MODULE_MAIN_NAME starting" >> "$LOG_DIR"/"$MAIN_LOG"

Reply:
works for me :)